### PR TITLE
[Merged by Bors] - feat(Complex/UpperHalfPlane): add vertical strips

### PR DESCRIPTION
**Pull request by CBirkbeck**
We define the vertical strips that are needed for proving Eisenstein series are modular forms #10377 . We also add the definition of sqrt{-1} as an elements of the upper half plane. Note that this is no longer needed for the modular forms PRs but its good to have.

Sorry about the typo in the PR name, its not my day!

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by loefflerd** [comment]
Since the vertical-strip stuff doesn't use the metric any more, it probably shouldn't be in `UpperHalfPlane/Metric.lean`.

**Comment by CBirkbeck** [comment]
> Since the vertical-strip stuff doesn't use the metric any more, it probably shouldn't be in `UpperHalfPlane/Metric.lean`.

Good point, I'm not quite sure where to place it, I guess it could go into topology?

**Review by grunweg — APPROVED** [review]
Looks straightforward; I just have three minor comments. Thanks for your contribution!

File: Mathlib/Analysis/Complex/UpperHalfPlane/Topology.lean [hunk 1]
```diff
@@ -74,4 +74,22 @@ instance : NoncompactSpace ℍ := by
...
+def verticalStrip (A B : ℝ) := {z : ℍ | |z.re| ≤ A ∧ B ≤ z.im}
+
+theorem strip_mem_iff (A B : ℝ) (z : ℍ) : z ∈ verticalStrip A B ↔ |z.re| ≤ A ∧ B ≤ z.im :=
+  Iff.rfl
+
+lemma subset_strip_of_isCompact {K : Set ℍ} (hK : IsCompact K) :
+    ∃ A B : ℝ, 0 < B ∧ K ⊆ verticalStrip A B := by
+  rcases K.eq_empty_or_nonempty with rfl | hne
+  · exact ⟨1, 1, Real.zero_lt_one, empty_subset _⟩
+  obtain ⟨u, _, hu⟩ := hK.exists_isMinOn hne continuous_im.continuousOn
```
[the remark this answers is not in the document]
Comment (CBirkbeck) [reply Mathlib/Analysis/Complex/UpperHalfPlane/Topology.lean:90]: I'm not sure I understand, you mean just swap what u and v are defined as?

File: Mathlib/Analysis/Complex/UpperHalfPlane/Topology.lean [hunk 2]
```diff
@@ -87,9 +87,9 @@ lemma subset_strip_of_isCompact {K : Set ℍ} (hK : IsCompact K) :
     ∃ A B : ℝ, 0 < B ∧ K ⊆ verticalStrip A B := by
   rcases K.eq_empty_or_nonempty with rfl | hne
   · exact ⟨1, 1, Real.zero_lt_one, empty_subset _⟩
-  obtain ⟨u, _, hu⟩ := hK.exists_isMinOn hne continuous_im.continuousOn
-  obtain ⟨v, _, hv⟩ := hK.exists_isMaxOn hne (_root_.continuous_abs.comp continuous_re).continuousOn
-  exact ⟨|re v|, im u, u.im_pos, fun k hk ↦ ⟨isMaxOn_iff.mp hv _ hk, isMinOn_iff.mp hu _ hk⟩⟩
+  obtain ⟨v, _, hv⟩ := hK.exists_isMinOn hne continuous_im.continuousOn
```
Comment (grunweg) [inline Mathlib/Analysis/Complex/UpperHalfPlane/Topology.lean:90]: If you care, you could also switch these two lines. (But this is really nano-comment territory.)

**Comment by loefflerd** [comment]
I think for consistency with mathlib's naming conventions, `strip_mem_iff` should probably be `mem_verticalStrip_iff`, and similarly `subset_verticalStrip_of_isCompact` for the other one. Otherwise looks good, I'd be happy to recommend this for merging modulo that change.

**Review by loefflerd — APPROVED** [review]
LGTM, thanks!

maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**